File: Leetcode/python/sol999.py
# https://leetcode.com/problems/available-captures-for-rook/

class Solution:
    def find_rook_pos(self, board: List[List[str]]) -> (int, int):
        size = len(board)
        for row in range(size):
            for col in range(size):
                if board[row][col] == "R":
                    return (row, col)

    def get_row_col(self, board, r_idx, c_idx) -> (List[int], List[int]):
        col = []
        for row in board:
            col.append(row[c_idx])
        return board[r_idx], col

    def check_pawn(self, arr: list, start: int):
        count = 0
        # left
        for i in range(start, len(arr)):
            if arr[i] == 'p':
                count += 1
                break
            elif arr[i] == 'B':
                break
        # right
        for i in range(start, -1, -1):
            if arr[i] == 'p':
                count += 1
                break
            elif arr[i] == 'B':
                break

        return count

    def numRookCaptures(self, board: List[List[str]]) -> int:
        row_idx, col_idx = self.find_rook_pos(board)
        row, col = self.get_row_col(board, row_idx, col_idx)
        return self.check_pawn(row, row_idx) + self.check_pawn(col, col_idx)

File: Leetcode/cpp/sol999.cpp
// https://leetcode.com/problems/available-captures-for-rook/

class Solution {
public:
    pair<int, int> find_rook(vector<vector<char>>& board){
        int row, col = 0;
        for(int i = 0; i<board.size(); i++){
            for(int j = 0; j<board[0].size(); j++){
                if(board[i][j] == 'R'){
                    row = i;
                    col = j;
                }
            }
        }
        return {row, col};
    }

    int numRookCaptures(vector<vector<char>>& board) {
        pair<int, int> pos = find_rook(board);
        vector<pair<int, int>> directions = {{-1,0},{1,0},{0,-1},{0,1}};
        int count  = 0;
        for(auto dir: directions){
            int row = dir.first + pos.first;
            int col = dir.second + pos.second;
            while(row < 8 && 0 <= row && col < 8 && 0 <= col){
                if(board[row][col] == 'p'){
                    count++;
                    break;
                }
                if(board[row][col] == 'B'){
                    break;
                }
                row+= dir.first;
                col+= dir.second;
            }

        }
        return count;
    }
};